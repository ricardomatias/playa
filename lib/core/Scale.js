import HarmonyBase from './HarmonyBase';
import Note from './Note';
import { SCALES } from './constants';
import assignOctaves from '../utils/assignOctaves';
import { interval as intervalTool } from '../tools';
import NoteType from './types';

/**
 * Defines a Scale
 * @typedef {Object} Scale
 *
 * @class Scale
 */
class Scale extends HarmonyBase {
	/**
	 * Creates an instance of Scale.
	 * @param {Note|String} note A note
	 * @param {String} type the type of scale to create
	 * @param {Object} [opts]
	 * @param {String} [opts.noteType] the note's value type
	 * @param {Array<Number>} [opts.octaves] [starting, number of octaves] range of octaves to map notes to
	 * @memberof Scale
	 */
	constructor(note, type, { noteType, octaves } = {}) {
		super(note, type, noteType);

		this._intervals = SCALES[type];
		this._octaves = octaves || [ 4, 1 ];

		if (!this._intervals) {
			throw new Error('That scale is not defined! Try another one.');
		}

		this._notes = this._createScale() || [];

		if (this._noteType === NoteType.MIDI || this._noteType === NoteType.FREQ) {
			this._notes = assignOctaves(this._notes, this._octaves, 'scale');
		}
	}

	/**
	 * Creates a chromatic scale
	 *
	 * @param {String} rootNote
	 * @param {Boolean} withFlats
	 * @return {Array<Note>} Chromatic notes
	 * @memberof Scale
	 */
	_createChromaticScale(rootNote, withFlats) {
		const chromaticNotes = [];

		let nextNote = rootNote;

		for (let index = -1; index < 12; index++) {
			let note = nextNote;

			if (note.isSharp() &&
				(rootNote.isFlat() || withFlats)) {
				note = new Note(note.e);
			}

			nextNote = note.next;

			chromaticNotes.push(note);
		}

		return chromaticNotes;
	}

	/**
	 * Creates a scale
	 *
	 * @return {Array<Note>} The defined scale notes
	 * @memberof Scale
	 */
	_createScale() {
		const notes = [];

		const rootNote = new Note(this._root);
		const scaleIntervals = this._intervals.split(' ');

		const chromaticNotes = this._createChromaticScale(rootNote);

		if (chromaticNotes.length === this._intervals.length) {
			return chromaticNotes;
		}

		for (let index = 0; index < scaleIntervals.length; index++) {
			const interval = scaleIntervals[index];
			const semitones = intervalTool.semitones(interval);

			if (semitones < chromaticNotes.length) {
				notes[index] = chromaticNotes[semitones];
			}
		}

		return notes;
	}
}

Scale.CHROMATIC = 'CHROMATIC';
Scale.LYDIAN = 'LYDIAN';
Scale.MAJOR = 'MAJOR';
Scale.IONIAN = 'IONIAN';
Scale.MIXOLYDIAN = 'MIXOLYDIAN';
Scale.MINOR = 'MINOR';
Scale.AEOLIAN = 'AEOLIAN';
Scale.DORIAN = 'DORIAN';
Scale.PHRYGIAN = 'PHRYGIAN';
Scale.LOCRIAN = 'LOCRIAN';
Scale.MAJOR_PENTATONIC = 'MAJOR_PENTATONIC';
Scale.MINOR_PENTATONIC = 'MINOR_PENTATONIC';
Scale.EGYPTIAN = 'EGYPTIAN';
Scale.MELODIC_MINOR = 'MELODIC_MINOR';
Scale.ALTERED = 'ALTERED';
Scale.HARMONIC_MINOR = 'HARMONIC_MINOR';

export default Scale;
