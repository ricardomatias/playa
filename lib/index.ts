// export * from './constants';
// export * from './composition';
// export * from './tools';
// export * from './analysis';

export { Note } from './core/Note';
// export { Chord } from './core/Chord';
export { Scale } from './core/Scale';
export { Key } from './core/Key';
// export * from './core/Time';

// export * from './core/Event';
// export * from './core/ChordEvent';
// export * from './core/NoteEvent';
