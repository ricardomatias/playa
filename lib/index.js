import * as Constants from './constants';
import * as Core from './core';
import * as Functional from './functional';
import * as Tools from './tools';

export {
	Core,
	Constants,
	Functional,
	Tools,
};
