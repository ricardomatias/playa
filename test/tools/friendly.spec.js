import friendly, { orderNotes, rankIntervals, rankScales } from '../../lib/tools/friendly';
import Random from '../../lib/tools/random';
import Scale from '../../lib/core/Scale';
import * as R from 'ramda';
import { stripOctave } from '../../lib/utils';


const Cmaj = new Scale('C', Scale.MAJOR);
const Abmaj = new Scale('Ab', Scale.MAJOR);
const DSharpLoc = new Scale('D#', Scale.LOCRIAN);
const BbEgyptian = new Scale('Bb', Scale.EGYPTIAN);

const SCALES_ARRAY = Array.from(Scale.SCALES.values());

describe('A Friendly test suite', () => {
	describe('classified list of neighboring scales', () => {
		it('should return [ `A`, `C#`, `G`, `B` ] ', () => {
			// given
			Random.setSeed('test');

			const neighbours = friendly([ 'A', 'C#', 'G', 'B' ]);

			expect(R.head(neighbours)).toEqual({
				scale: '1P 2M 3M 4P 5P 6M 7m', match: 1, type: 'MIXOLYDIAN', note: 'A',
			});

			expect(neighbours).toMatchSnapshot();
		});

		it('should return [ `F#`, `C#`, `D`, `D`, `E`, `C` ]', () => {
			// given
			Random.setSeed('test');

			const neighbours = friendly([ 'F#', 'C#', 'D', 'D', 'E', 'C' ]);

			expect(R.head(neighbours)).toEqual({
				scale: '1P 2M 3M 4P 5P 6M 7m', match: 0.75, type: 'MIXOLYDIAN', note: 'D',
			});

			expect(neighbours).toMatchSnapshot();
		});

		it('should return [ `B`, `D`, `Eb`, `F`, `Ab` ]', () => {
			// given
			Random.setSeed('test');

			const neighbours = friendly([ 'B', 'D', 'Eb', 'F', 'Ab' ]);

			expect(R.head(neighbours)).toEqual({
				scale: '1P 2M 3M 4P 5P 6M 7M', match: 0.5, note: 'Ab', type: 'IONIAN',
			});

			expect(neighbours).toMatchSnapshot();
		});

		it('should still return neighbours when given just 2 notes', () => {
			// given
			Random.setSeed('test');

			// when
			const neighbours = friendly([ 'C#', 'F#' ]);

			// then
			expect(R.head(neighbours)).toEqual({ scale: '1P 2M 3M 4A 5P 6M 7M', match: 1, note: 'F#', type: 'LYDIAN' });
			expect(neighbours).toMatchSnapshot();
		});

		it('should return an empty array when given 1 note', () => {
			// given
			Random.setSeed('test');

			// when
			const neighbours = friendly([ 'C#' ]);

			// then
			expect(neighbours).toEqual([]);
		});
	});

	describe('#rank', () => {
		it('should return an ordered list of intervals', () => {
			const ranking = rankIntervals(SCALES_ARRAY);

			expect(ranking).toEqual([
				'4A', '5d', '2m', '7M', '3M', '6m',
				'6M', '3m', '7m', '2M', '4P', '5P',
			]);
		});

		it('should return an ordered list of scales based on intervals ranking', () => {
			const intervalsRanking = rankIntervals(SCALES_ARRAY);

			const ranking = rankScales(intervalsRanking);

			expect(ranking['4A']).toEqual([ '1P 2M 3M 4A 5P 6M 7M' ]);
			expect(ranking['7m']).toHaveLength(9);
			expect(ranking['5P']).toHaveLength(13);
			expect(ranking).toMatchSnapshot();
		});
	});

	describe('#orderNotes', () => {
		beforeAll(() => Random.setSeed('FRIENDLY'));

		it('should order notes C MAJ', () => {
			const rndFn = () => (Random.int(1, -1));

			const scrambledNotes = R.sort(rndFn, Cmaj.string);

			const neighbours = orderNotes(scrambledNotes);

			expect(neighbours).toEqual(Cmaj.string.map(stripOctave));
		});

		it('should order notes Ab MAJ', () => {
			const rndFn = () => (Random.int(1, -1));

			const scrambledNotes = R.sort(rndFn, Abmaj.string);
			const neighbours = orderNotes(scrambledNotes);

			expect(neighbours).toEqual([
				'C', 'Db', 'Eb', 'F', 'G', 'Ab', 'Bb',
			]);
		});

		it('should order notes D# LOCRIAN', () => {
			const rndFn = () => (Random.int(1, -1));

			const scrambledNotes = R.sort(rndFn, DSharpLoc.string);

			expect(scrambledNotes).toEqual([ 'E3', 'G#3', 'F#3', 'D#3', 'B3', 'C#4', 'A3' ]);

			const neighbours = orderNotes(scrambledNotes);

			expect(neighbours).toEqual([
				'C#', 'D#', 'E', 'F#', 'G#', 'A', 'B',
			]);
		});

		it('should order notes Bb EGYPTIAN', () => {
			const rndFn = () => (Random.int(1, -1));

			const scrambledNotes = R.sort(rndFn, BbEgyptian.string);
			const neighbours = orderNotes(scrambledNotes);

			expect(neighbours).toEqual([
				'C', 'Eb', 'F', 'Ab', 'Bb',
			]);
		});
	});
});
