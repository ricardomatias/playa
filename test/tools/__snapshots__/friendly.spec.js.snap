// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`A Friendly test suite #rank should return an ordered list of scales based on intervals ranking 1`] = `
Object {
  "2M": Array [
    "1P 2M 3M 4A 5P 6M 7M",
    "1P 2M 3M 4P 5P 6M 7M",
    "1P 2M 3M 4P 5P 6M 7M",
    "1P 2M 3M 4P 5P 6M 7m",
    "1P 2M 3m 4P 5P 6m 7m",
    "1P 2M 3m 4P 5P 6m 7m",
    "1P 2M 3m 4P 5P 6M 7m",
    "1P 2M 3M 5P 6M",
    "1P 2M 4P 5P 7m",
    "1P 2M 3m 4P 5P 6M 7M",
    "1P 2M 3m 4P 5P 6m 7M",
  ],
  "2m": Array [
    "1P 2m 3m 4P 5P 6m 7m",
    "1P 2m 3m 4P 5d 6m 7m",
    "1P 2m 3m 3M 5d 6m 7m",
  ],
  "3M": Array [
    "1P 2M 3M 4A 5P 6M 7M",
    "1P 2M 3M 4P 5P 6M 7M",
    "1P 2M 3M 4P 5P 6M 7M",
    "1P 2M 3M 4P 5P 6M 7m",
    "1P 2M 3M 5P 6M",
    "1P 2m 3m 3M 5d 6m 7m",
  ],
  "3m": Array [
    "1P 2M 3m 4P 5P 6m 7m",
    "1P 2M 3m 4P 5P 6m 7m",
    "1P 2M 3m 4P 5P 6M 7m",
    "1P 2m 3m 4P 5P 6m 7m",
    "1P 2m 3m 4P 5d 6m 7m",
    "1P 3m 4P 5P 7m",
    "1P 2M 3m 4P 5P 6M 7M",
    "1P 2m 3m 3M 5d 6m 7m",
    "1P 2M 3m 4P 5P 6m 7M",
  ],
  "4A": Array [
    "1P 2M 3M 4A 5P 6M 7M",
  ],
  "4P": Array [
    "1P 2M 3M 4P 5P 6M 7M",
    "1P 2M 3M 4P 5P 6M 7M",
    "1P 2M 3M 4P 5P 6M 7m",
    "1P 2M 3m 4P 5P 6m 7m",
    "1P 2M 3m 4P 5P 6m 7m",
    "1P 2M 3m 4P 5P 6M 7m",
    "1P 2m 3m 4P 5P 6m 7m",
    "1P 2m 3m 4P 5d 6m 7m",
    "1P 3m 4P 5P 7m",
    "1P 2M 4P 5P 7m",
    "1P 2M 3m 4P 5P 6M 7M",
    "1P 2M 3m 4P 5P 6m 7M",
  ],
  "5P": Array [
    "1P 2M 3M 4A 5P 6M 7M",
    "1P 2M 3M 4P 5P 6M 7M",
    "1P 2M 3M 4P 5P 6M 7M",
    "1P 2M 3M 4P 5P 6M 7m",
    "1P 2M 3m 4P 5P 6m 7m",
    "1P 2M 3m 4P 5P 6m 7m",
    "1P 2M 3m 4P 5P 6M 7m",
    "1P 2m 3m 4P 5P 6m 7m",
    "1P 2M 3M 5P 6M",
    "1P 3m 4P 5P 7m",
    "1P 2M 4P 5P 7m",
    "1P 2M 3m 4P 5P 6M 7M",
    "1P 2M 3m 4P 5P 6m 7M",
  ],
  "5d": Array [
    "1P 2m 3m 4P 5d 6m 7m",
    "1P 2m 3m 3M 5d 6m 7m",
  ],
  "6M": Array [
    "1P 2M 3M 4A 5P 6M 7M",
    "1P 2M 3M 4P 5P 6M 7M",
    "1P 2M 3M 4P 5P 6M 7M",
    "1P 2M 3M 4P 5P 6M 7m",
    "1P 2M 3m 4P 5P 6M 7m",
    "1P 2M 3M 5P 6M",
    "1P 2M 3m 4P 5P 6M 7M",
  ],
  "6m": Array [
    "1P 2M 3m 4P 5P 6m 7m",
    "1P 2M 3m 4P 5P 6m 7m",
    "1P 2m 3m 4P 5P 6m 7m",
    "1P 2m 3m 4P 5d 6m 7m",
    "1P 2m 3m 3M 5d 6m 7m",
    "1P 2M 3m 4P 5P 6m 7M",
  ],
  "7M": Array [
    "1P 2M 3M 4A 5P 6M 7M",
    "1P 2M 3M 4P 5P 6M 7M",
    "1P 2M 3M 4P 5P 6M 7M",
    "1P 2M 3m 4P 5P 6M 7M",
    "1P 2M 3m 4P 5P 6m 7M",
  ],
  "7m": Array [
    "1P 2M 3M 4P 5P 6M 7m",
    "1P 2M 3m 4P 5P 6m 7m",
    "1P 2M 3m 4P 5P 6m 7m",
    "1P 2M 3m 4P 5P 6M 7m",
    "1P 2m 3m 4P 5P 6m 7m",
    "1P 2m 3m 4P 5d 6m 7m",
    "1P 3m 4P 5P 7m",
    "1P 2M 4P 5P 7m",
    "1P 2m 3m 3M 5d 6m 7m",
  ],
}
`;

exports[`A Friendly test suite classified list of neighboring scales should return [ \`A\`, \`C#\`, \`G\`, \`B\` ]  1`] = `
Array [
  Object {
    "match": 1,
    "note": "A",
    "scale": "1P 2M 3M 4P 5P 6M 7m",
    "type": "MIXOLYDIAN",
  },
  Object {
    "match": 1,
    "note": "G",
    "scale": "1P 2M 3M 4A 5P 6M 7M",
    "type": "LYDIAN",
  },
  Object {
    "match": 0.3333333333333333,
    "note": "A",
    "scale": "1P 2M 3m 4P 5P 6M 7M",
    "type": "MELODIC_MINOR",
  },
  Object {
    "match": 0.3333333333333333,
    "note": "A",
    "scale": "1P 2M 3m 4P 5P 6m 7M",
    "type": "HARMONIC_MINOR",
  },
  Object {
    "match": 0.3333333333333333,
    "note": "A",
    "scale": "1P 2m 3m 4P 5P 6m 7m",
    "type": "PHRYGIAN",
  },
  Object {
    "match": 0.3333333333333333,
    "note": "A",
    "scale": "1P 2m 3m 4P 5d 6m 7m",
    "type": "LOCRIAN",
  },
  Object {
    "match": 0.3333333333333333,
    "note": "A",
    "scale": "1P 3m 4P 5P 7m",
    "type": "MINOR_PENTATONIC",
  },
  Object {
    "match": 0.3333333333333333,
    "note": "B",
    "scale": "1P 2M 3M 4A 5P 6M 7M",
    "type": "LYDIAN",
  },
  Object {
    "match": 0.3333333333333333,
    "note": "B",
    "scale": "1P 2M 3M 4P 5P 6M 7M",
    "type": "IONIAN",
  },
  Object {
    "match": 0.3333333333333333,
    "note": "B",
    "scale": "1P 2M 3M 5P 6M",
    "type": "MAJOR_PENTATONIC",
  },
  Object {
    "match": 0.3333333333333333,
    "note": "B",
    "scale": "1P 2M 3m 4P 5P 6M 7M",
    "type": "MELODIC_MINOR",
  },
  Object {
    "match": 0.3333333333333333,
    "note": "B",
    "scale": "1P 2M 3m 4P 5P 6m 7M",
    "type": "HARMONIC_MINOR",
  },
  Object {
    "match": 0.3333333333333333,
    "note": "B",
    "scale": "1P 2m 3m 4P 5P 6m 7m",
    "type": "PHRYGIAN",
  },
  Object {
    "match": 0.3333333333333333,
    "note": "B",
    "scale": "1P 2m 3m 4P 5d 6m 7m",
    "type": "LOCRIAN",
  },
  Object {
    "match": 0.3333333333333333,
    "note": "B",
    "scale": "1P 3m 4P 5P 7m",
    "type": "MINOR_PENTATONIC",
  },
  Object {
    "match": 0.3333333333333333,
    "note": "B",
    "scale": "1P 2m 3m 3M 5d 6m 7m",
    "type": "ALTERED",
  },
  Object {
    "match": 0.3333333333333333,
    "note": "G",
    "scale": "1P 2M 3m 4P 5P 6m 7m",
    "type": "AEOLIAN",
  },
  Object {
    "match": 0.3333333333333333,
    "note": "G",
    "scale": "1P 2M 3m 4P 5P 6M 7m",
    "type": "DORIAN",
  },
  Object {
    "match": 0.3333333333333333,
    "note": "G",
    "scale": "1P 2M 4P 5P 7m",
    "type": "EGYPTIAN",
  },
  Object {
    "match": 0.3333333333333333,
    "note": "G",
    "scale": "1P 2M 3m 4P 5P 6M 7M",
    "type": "MELODIC_MINOR",
  },
  Object {
    "match": 0.3333333333333333,
    "note": "G",
    "scale": "1P 2M 3m 4P 5P 6m 7M",
    "type": "HARMONIC_MINOR",
  },
  Object {
    "match": 0.3333333333333333,
    "note": "G",
    "scale": "1P 2m 3m 3M 5d 6m 7m",
    "type": "ALTERED",
  },
  Object {
    "match": 0.3333333333333333,
    "note": "C#",
    "scale": "1P 2M 3M 4A 5P 6M 7M",
    "type": "LYDIAN",
  },
  Object {
    "match": 0.3333333333333333,
    "note": "C#",
    "scale": "1P 2M 3M 4P 5P 6M 7m",
    "type": "MIXOLYDIAN",
  },
  Object {
    "match": 0.3333333333333333,
    "note": "C#",
    "scale": "1P 2M 3m 4P 5P 6m 7m",
    "type": "AEOLIAN",
  },
  Object {
    "match": 0.3333333333333333,
    "note": "C#",
    "scale": "1P 2M 3m 4P 5P 6M 7m",
    "type": "DORIAN",
  },
  Object {
    "match": 0.3333333333333333,
    "note": "C#",
    "scale": "1P 2m 3m 4P 5P 6m 7m",
    "type": "PHRYGIAN",
  },
  Object {
    "match": 0.3333333333333333,
    "note": "C#",
    "scale": "1P 2m 3m 4P 5d 6m 7m",
    "type": "LOCRIAN",
  },
  Object {
    "match": 0.3333333333333333,
    "note": "C#",
    "scale": "1P 3m 4P 5P 7m",
    "type": "MINOR_PENTATONIC",
  },
  Object {
    "match": 0.3333333333333333,
    "note": "C#",
    "scale": "1P 2M 4P 5P 7m",
    "type": "EGYPTIAN",
  },
  Object {
    "match": 0.3333333333333333,
    "note": "C#",
    "scale": "1P 2m 3m 3M 5d 6m 7m",
    "type": "ALTERED",
  },
]
`;

exports[`A Friendly test suite classified list of neighboring scales should return [ \`B\`, \`D\`, \`Eb\`, \`F\`, \`Ab\` ] 1`] = `
Array [
  Object {
    "match": 0.5,
    "note": "Ab",
    "scale": "1P 2M 3M 4P 5P 6M 7M",
    "type": "IONIAN",
  },
  Object {
    "match": 0.5,
    "note": "Ab",
    "scale": "1P 2M 3M 4P 5P 6M 7m",
    "type": "MIXOLYDIAN",
  },
  Object {
    "match": 0.5,
    "note": "Ab",
    "scale": "1P 2M 3M 5P 6M",
    "type": "MAJOR_PENTATONIC",
  },
  Object {
    "match": 0.5,
    "note": "F",
    "scale": "1P 2M 3M 4P 5P 6M 7m",
    "type": "MIXOLYDIAN",
  },
  Object {
    "match": 0.5,
    "note": "F",
    "scale": "1P 2M 3M 4P 5P 6M 7m",
    "type": "MIXOLYDIAN",
  },
  Object {
    "match": 0.5,
    "note": "D",
    "scale": "1P 2M 3M 4A 5P 6M 7M",
    "type": "LYDIAN",
  },
  Object {
    "match": 0.25,
    "note": "Ab",
    "scale": "1P 2m 3m 4P 5d 6m 7m",
    "type": "LOCRIAN",
  },
  Object {
    "match": 0.25,
    "note": "Ab",
    "scale": "1P 2m 3m 3M 5d 6m 7m",
    "type": "ALTERED",
  },
  Object {
    "match": 0.25,
    "note": "Ab",
    "scale": "1P 2M 4P 5P 7m",
    "type": "EGYPTIAN",
  },
  Object {
    "match": 0.25,
    "note": "F",
    "scale": "1P 2M 3m 4P 5P 6m 7M",
    "type": "HARMONIC_MINOR",
  },
  Object {
    "match": 0.25,
    "note": "F",
    "scale": "1P 2M 3M 4P 5P 6M 7M",
    "type": "IONIAN",
  },
  Object {
    "match": 0.25,
    "note": "F",
    "scale": "1P 2M 3M 5P 6M",
    "type": "MAJOR_PENTATONIC",
  },
  Object {
    "match": 0.25,
    "note": "F",
    "scale": "1P 2M 4P 5P 7m",
    "type": "EGYPTIAN",
  },
  Object {
    "match": 0.25,
    "note": "F",
    "scale": "1P 2M 3m 4P 5P 6m 7M",
    "type": "HARMONIC_MINOR",
  },
  Object {
    "match": 0.25,
    "note": "F",
    "scale": "1P 2M 3M 4P 5P 6M 7M",
    "type": "IONIAN",
  },
  Object {
    "match": 0.25,
    "note": "F",
    "scale": "1P 2M 3M 5P 6M",
    "type": "MAJOR_PENTATONIC",
  },
  Object {
    "match": 0.25,
    "note": "F",
    "scale": "1P 2M 4P 5P 7m",
    "type": "EGYPTIAN",
  },
  Object {
    "match": 0.25,
    "note": "B",
    "scale": "1P 2M 3m 4P 5P 6m 7m",
    "type": "AEOLIAN",
  },
  Object {
    "match": 0.25,
    "note": "B",
    "scale": "1P 2m 3m 4P 5P 6m 7m",
    "type": "PHRYGIAN",
  },
  Object {
    "match": 0.25,
    "note": "B",
    "scale": "1P 2m 3m 4P 5d 6m 7m",
    "type": "LOCRIAN",
  },
  Object {
    "match": 0.25,
    "note": "B",
    "scale": "1P 3m 4P 5P 7m",
    "type": "MINOR_PENTATONIC",
  },
  Object {
    "match": 0.25,
    "note": "B",
    "scale": "1P 2M 3m 4P 5P 6m 7M",
    "type": "HARMONIC_MINOR",
  },
  Object {
    "match": 0.25,
    "note": "D",
    "scale": "1P 2M 3m 4P 5P 6m 7m",
    "type": "AEOLIAN",
  },
  Object {
    "match": 0.25,
    "note": "D",
    "scale": "1P 3m 4P 5P 7m",
    "type": "MINOR_PENTATONIC",
  },
  Object {
    "match": 0.25,
    "note": "D",
    "scale": "1P 2M 3m 4P 5P 6m 7M",
    "type": "HARMONIC_MINOR",
  },
  Object {
    "match": 0.25,
    "note": "D",
    "scale": "1P 2M 3M 4P 5P 6M 7M",
    "type": "IONIAN",
  },
  Object {
    "match": 0.25,
    "note": "D",
    "scale": "1P 2M 3M 4P 5P 6M 7m",
    "type": "MIXOLYDIAN",
  },
  Object {
    "match": 0.25,
    "note": "D",
    "scale": "1P 2M 3M 5P 6M",
    "type": "MAJOR_PENTATONIC",
  },
]
`;

exports[`A Friendly test suite classified list of neighboring scales should return [ \`F#\`, \`C#\`, \`D\`, \`D\`, \`E\`, \`C\` ] 1`] = `
Array [
  Object {
    "match": 0.75,
    "note": "D",
    "scale": "1P 2M 3M 4P 5P 6M 7m",
    "type": "MIXOLYDIAN",
  },
  Object {
    "match": 0.75,
    "note": "D",
    "scale": "1P 2M 3M 4A 5P 6M 7M",
    "type": "LYDIAN",
  },
  Object {
    "match": 0.75,
    "note": "D",
    "scale": "1P 2M 3M 4P 5P 6M 7M",
    "type": "IONIAN",
  },
  Object {
    "match": 0.75,
    "note": "E",
    "scale": "1P 2M 3M 4P 5P 6M 7m",
    "type": "MIXOLYDIAN",
  },
  Object {
    "match": 0.75,
    "note": "E",
    "scale": "1P 2M 3m 4P 5P 6M 7m",
    "type": "DORIAN",
  },
  Object {
    "match": 0.75,
    "note": "E",
    "scale": "1P 2M 3M 4P 5P 6M 7m",
    "type": "MIXOLYDIAN",
  },
  Object {
    "match": 0.75,
    "note": "E",
    "scale": "1P 2M 3m 4P 5P 6M 7m",
    "type": "DORIAN",
  },
  Object {
    "match": 0.25,
    "note": "D",
    "scale": "1P 2m 3m 4P 5P 6m 7m",
    "type": "PHRYGIAN",
  },
  Object {
    "match": 0.25,
    "note": "D",
    "scale": "1P 2m 3m 4P 5d 6m 7m",
    "type": "LOCRIAN",
  },
  Object {
    "match": 0.25,
    "note": "D",
    "scale": "1P 3m 4P 5P 7m",
    "type": "MINOR_PENTATONIC",
  },
  Object {
    "match": 0.25,
    "note": "E",
    "scale": "1P 2M 3m 4P 5P 6m 7M",
    "type": "HARMONIC_MINOR",
  },
  Object {
    "match": 0.25,
    "note": "E",
    "scale": "1P 2m 3m 4P 5P 6m 7m",
    "type": "PHRYGIAN",
  },
  Object {
    "match": 0.25,
    "note": "E",
    "scale": "1P 2m 3m 4P 5d 6m 7m",
    "type": "LOCRIAN",
  },
  Object {
    "match": 0.25,
    "note": "E",
    "scale": "1P 3m 4P 5P 7m",
    "type": "MINOR_PENTATONIC",
  },
  Object {
    "match": 0.25,
    "note": "E",
    "scale": "1P 2m 3m 3M 5d 6m 7m",
    "type": "ALTERED",
  },
  Object {
    "match": 0.25,
    "note": "E",
    "scale": "1P 2M 3m 4P 5P 6m 7M",
    "type": "HARMONIC_MINOR",
  },
  Object {
    "match": 0.25,
    "note": "E",
    "scale": "1P 2m 3m 4P 5P 6m 7m",
    "type": "PHRYGIAN",
  },
  Object {
    "match": 0.25,
    "note": "E",
    "scale": "1P 2m 3m 4P 5d 6m 7m",
    "type": "LOCRIAN",
  },
  Object {
    "match": 0.25,
    "note": "E",
    "scale": "1P 3m 4P 5P 7m",
    "type": "MINOR_PENTATONIC",
  },
  Object {
    "match": 0.25,
    "note": "E",
    "scale": "1P 2m 3m 3M 5d 6m 7m",
    "type": "ALTERED",
  },
  Object {
    "match": 0.25,
    "note": "F#",
    "scale": "1P 2M 3M 4P 5P 6M 7M",
    "type": "IONIAN",
  },
  Object {
    "match": 0.25,
    "note": "F#",
    "scale": "1P 2M 3M 5P 6M",
    "type": "MAJOR_PENTATONIC",
  },
  Object {
    "match": 0.25,
    "note": "F#",
    "scale": "1P 2M 3m 4P 5P 6M 7M",
    "type": "MELODIC_MINOR",
  },
  Object {
    "match": 0.25,
    "note": "F#",
    "scale": "1P 2M 3m 4P 5P 6m 7M",
    "type": "HARMONIC_MINOR",
  },
  Object {
    "match": 0.25,
    "note": "F#",
    "scale": "1P 2m 3m 4P 5d 6m 7m",
    "type": "LOCRIAN",
  },
  Object {
    "match": 0.25,
    "note": "F#",
    "scale": "1P 2m 3m 3M 5d 6m 7m",
    "type": "ALTERED",
  },
  Object {
    "match": 0.25,
    "note": "C",
    "scale": "1P 2m 3m 4P 5P 6m 7m",
    "type": "PHRYGIAN",
  },
  Object {
    "match": 0.25,
    "note": "C",
    "scale": "1P 2m 3m 4P 5d 6m 7m",
    "type": "LOCRIAN",
  },
  Object {
    "match": 0.25,
    "note": "C",
    "scale": "1P 2M 3m 4P 5P 6m 7m",
    "type": "AEOLIAN",
  },
  Object {
    "match": 0.25,
    "note": "C",
    "scale": "1P 2M 3m 4P 5P 6M 7m",
    "type": "DORIAN",
  },
  Object {
    "match": 0.25,
    "note": "C",
    "scale": "1P 2M 4P 5P 7m",
    "type": "EGYPTIAN",
  },
  Object {
    "match": 0.25,
    "note": "C",
    "scale": "1P 2M 3m 4P 5P 6M 7M",
    "type": "MELODIC_MINOR",
  },
  Object {
    "match": 0.25,
    "note": "C",
    "scale": "1P 2M 3m 4P 5P 6m 7M",
    "type": "HARMONIC_MINOR",
  },
]
`;

exports[`A Friendly test suite classified list of neighboring scales should still return neighbours when given just 2 notes 1`] = `
Array [
  Object {
    "match": 1,
    "note": "F#",
    "scale": "1P 2M 3M 4A 5P 6M 7M",
    "type": "LYDIAN",
  },
  Object {
    "match": 1,
    "note": "F#",
    "scale": "1P 2M 3M 4P 5P 6M 7M",
    "type": "IONIAN",
  },
  Object {
    "match": 1,
    "note": "F#",
    "scale": "1P 2M 3M 4P 5P 6M 7M",
    "type": "IONIAN",
  },
  Object {
    "match": 1,
    "note": "F#",
    "scale": "1P 2M 3M 4P 5P 6M 7m",
    "type": "MIXOLYDIAN",
  },
  Object {
    "match": 1,
    "note": "F#",
    "scale": "1P 2M 3m 4P 5P 6m 7m",
    "type": "AEOLIAN",
  },
  Object {
    "match": 1,
    "note": "F#",
    "scale": "1P 2M 3m 4P 5P 6m 7m",
    "type": "AEOLIAN",
  },
  Object {
    "match": 1,
    "note": "F#",
    "scale": "1P 2M 3m 4P 5P 6M 7m",
    "type": "DORIAN",
  },
  Object {
    "match": 1,
    "note": "F#",
    "scale": "1P 2m 3m 4P 5P 6m 7m",
    "type": "PHRYGIAN",
  },
  Object {
    "match": 1,
    "note": "F#",
    "scale": "1P 2M 3M 5P 6M",
    "type": "MAJOR_PENTATONIC",
  },
  Object {
    "match": 1,
    "note": "F#",
    "scale": "1P 3m 4P 5P 7m",
    "type": "MINOR_PENTATONIC",
  },
  Object {
    "match": 1,
    "note": "F#",
    "scale": "1P 2M 4P 5P 7m",
    "type": "EGYPTIAN",
  },
  Object {
    "match": 1,
    "note": "F#",
    "scale": "1P 2M 3m 4P 5P 6M 7M",
    "type": "MELODIC_MINOR",
  },
  Object {
    "match": 1,
    "note": "F#",
    "scale": "1P 2M 3m 4P 5P 6m 7M",
    "type": "HARMONIC_MINOR",
  },
  Object {
    "match": 1,
    "note": "C#",
    "scale": "1P 2M 3M 4P 5P 6M 7M",
    "type": "IONIAN",
  },
  Object {
    "match": 1,
    "note": "C#",
    "scale": "1P 2M 3M 4P 5P 6M 7M",
    "type": "IONIAN",
  },
  Object {
    "match": 1,
    "note": "C#",
    "scale": "1P 2M 3M 4P 5P 6M 7m",
    "type": "MIXOLYDIAN",
  },
  Object {
    "match": 1,
    "note": "C#",
    "scale": "1P 2M 3m 4P 5P 6m 7m",
    "type": "AEOLIAN",
  },
  Object {
    "match": 1,
    "note": "C#",
    "scale": "1P 2M 3m 4P 5P 6m 7m",
    "type": "AEOLIAN",
  },
  Object {
    "match": 1,
    "note": "C#",
    "scale": "1P 2M 3m 4P 5P 6M 7m",
    "type": "DORIAN",
  },
  Object {
    "match": 1,
    "note": "C#",
    "scale": "1P 2m 3m 4P 5P 6m 7m",
    "type": "PHRYGIAN",
  },
  Object {
    "match": 1,
    "note": "C#",
    "scale": "1P 2m 3m 4P 5d 6m 7m",
    "type": "LOCRIAN",
  },
  Object {
    "match": 1,
    "note": "C#",
    "scale": "1P 3m 4P 5P 7m",
    "type": "MINOR_PENTATONIC",
  },
  Object {
    "match": 1,
    "note": "C#",
    "scale": "1P 2M 4P 5P 7m",
    "type": "EGYPTIAN",
  },
  Object {
    "match": 1,
    "note": "C#",
    "scale": "1P 2M 3m 4P 5P 6M 7M",
    "type": "MELODIC_MINOR",
  },
  Object {
    "match": 1,
    "note": "C#",
    "scale": "1P 2M 3m 4P 5P 6m 7M",
    "type": "HARMONIC_MINOR",
  },
]
`;
